### ISSUE TITLE
Let's not make the 'central, single entry-point' mistake again

### ISSUE THREAD

**Issue by lgrahl** [w3c/webtransport#180]
Happy new year to all of you!

I'm worried about the direction the API design is going. Let me explain.

I always thought WebTransport would define a low-level baseline for future web transports by outlining interface mixins making heavy use of byte streams using the streams API. These interface mixins could then be used, combined and extended by all sorts of transports. I for example made a proposal how this could look for SCTP (https://github.com/w3c/webtransport/issues/49). The beauty of this design is that the common baseline responsible for its composability are the byte streams of the streams API. Whether you are using HTTP/3, QUIC, SCTP or whatever else comes to your mind, it doesn't matter, you can strap a byte stream at it and that makes it composable.

Since then, datagrams have been added which added more complexity. However the `DatagramTransport` is also a mixin, so we are fine as long as this is not a requirement for future web transports.

Now, let's talk about the `WebTransport` interface. It is a central entrypoint, multiplexing protocols and I do not think we should follow that design. IMHO the pitfall of WebRTC API is the central entrypoint, the high-level `RTCPeerConnection`, responsible for much of WebRTCs complexity. Different transports have different properties for construction and use and we already have composability due to the byte streams and the mixins, so why mangle them all into a single entrypoint? What's the advantage in doing that?

Now, if `WebTransport` would not include `UnidirectionalStreamsTransport`, `BidirectionalStreamsTransport` and `DatagramTransport` and not have a constructor, which would then just define the mere baseline that any web transport must bring (e.g. the *ready* and *closed* promises), then I would be fine. A concrete `QuicTransport` could then include that and add further mixins or custom methods meeting the needs of QUIC, if needed. This also makes the standard future-proof because adding another transport just creates a new transport interface that does not automatically interfere with other transport protocols if a new property is needed.

AFAIK this change came naturally with the introduction of the URL since these are, by design, multiplexing many different things, including transport protocols. Let me make it clear: I'm not entirely against a high-level entrypoint to construct just *some* web transport if people really want that. However, I do think this should be added **on top of existing low-level building blocks**, if at all. I'm also not ignorant to the discussion you're having on the mailing list about load balancing and connection reuse. However, I do not see how this would conflict with low-level transport construction blocks.

**Comment by yutakahirano** [comment]
Is streams (a readable|writable stream, a pair of readable and writable streams or a bundle of them) enough for your abstraction?

**Comment by lgrahl** [comment]
I can't follow. Can you rephrase that question?

**Comment by yutakahirano** [comment]
> Whether you are using HTTP/3, QUIC, SCTP or whatever else comes to your mind, it doesn't matter, you can strap a byte stream at it and that makes it composable.

This sounds like you're not interested in protocol details and want to use a byte stream. Then, isn't byte stream the right abstraction for your application?

**Comment by lgrahl** [comment]
I do not have a particular application in mind here. This is not about my hypothetical application or use case but about the spec and its current lack of future-proofness due to the design of a generic, single entry-point with a URL. Let's take a quick look at it:

```js
[Exposed=(Window,Worker)]
interface WebTransport {
  constructor(USVString url, optional WebTransportOptions options = {});
  Promise<WebTransportStats> getStats();
  readonly attribute WebTransportState state;
  readonly attribute Promise<undefined> ready;
  readonly attribute Promise<WebTransportCloseInfo> closed;
  undefined close(optional WebTransportCloseInfo closeInfo
 = {});
  attribute EventHandler onstatechange;
};

WebTransport includes UnidirectionalStreamsTransport;
WebTransport includes BidirectionalStreamsTransport;
WebTransport includes DatagramTransport;
```

Sure, we can map QUIC and HTTP/3 to this design. But what do we do if there's a hypothetical QUICER transport that has another state which would require extending the `WebTransportState`? What would extending that interface mean to existing applications?

What do we do if there's a simple TCP-like transport which does not support datagrams? Should the datagram support be emulated, adding no advantage but additional complexity?

---

The alternative I'd prefer is to keep the mixins and have multiple entry-points. One for each transports. Because every transport is different.

```js
interface WebTransport<Stats, State, CloseInfo> {
  Promise<Stats> getStats();
  readonly attribute State state;
  readonly attribute Promise<undefined> ready;
  readonly attribute Promise<CloseInfo> closed;
  undefined close(optional WebTransportCloseInfo closeInfo
 = {});
  attribute EventHandler onstatechange;
}
```

This abstraction defines the mere baseline on what **every** WebTransport must be able to support. IIRC WebIDL has no concept of generics but it's simple: The structure always looks like this but e.g. states and close info may look different depending on the transport. (If this requires copy-pasting the interface, then so be it. You get the idea.)

```js
[Exposed=(Window,Worker)]
interface QuicTransport {
  constructor(USVString url, optional QuicTransportOptions options = {});
};

QuicTransport includes WebTransport<QuicStats, QuicState, QuicCloseInfo>
QuicTransport includes UnidirectionalStreamsTransport;
QuicTransport includes BidirectionalStreamsTransport;
QuicTransport includes DatagramTransport;
```

The QUIC transport now implements the mere baseline and interface mixins that it supports. Analogous to this the HTTP/3 transport:

```js
[Exposed=(Window,Worker)]
interface Http3Transport {
  constructor(USVString url, optional Http3TransportOptions options = {});
};

Http3Transport includes WebTransport<Http3Stats, Http3State, Http3CloseInfo>
Http3Transport includes UnidirectionalStreamsTransport;
Http3Transport includes BidirectionalStreamsTransport;
Http3Transport includes DatagramTransport;
```

 A hypothetical TLS transport, which does only have one bidirectional stream by design, could then look like this.

```js
[Exposed=(Window,Worker)]
interface TlsTransport {
  constructor(USVString url, optional TlsTransportOptions options = {});
};

TlsTransport includes WebTransport<QuicStats, QuicState, QuicCloseInfo>
TlsTransport includes BidirectionalStream
```

The advantage is that we can extend the WebTransport standard, as shown above, instead of inventing a new one like it was done in the past.

**Comment by jan-ivar** [comment]
> It is a central entrypoint, multiplexing protocols 

@lgrahl As of [last week](https://mailarchive.ietf.org/arch/msg/webtransport/9k27ZTwBlg9ncsZQXymg-jAfUOA/), HTTP/3 looks to be the sole UDP-based transport for WebTransport. As a result, we'll presumably deprecate the `quic-transport:` scheme, leaving `https:` as the lone URI scheme (modulo some ideas from @vasilvv I'll leave for him to propose).

Does that resolve your API concern?

**Comment by jan-ivar** [comment]
Spec is now focused on HTTP/3 (with HTTP/2 fallback). Mixins were removed, so any additions of other transports would need to be through other means than WebIDL mixins. Closing.
